### DISCUSSION TITLE
FleetAPI

### DISCUSSION THREAD

**Discussion by jasonacox** [jasonacox/pypowerwall#71] [General]
# Tesla FleetAPI for Powerwall

I have added a new `fleetapi.py` tool and python class: https://github.com/jasonacox/pypowerwall/tree/main/tools/fleetapi

[FleetAPI](https://developer.tesla.com/docs/fleet-api#overview) is the official Tesla API RESTful data and command service for Tesla products, including Powerwalls. Users can interact with their own devices, or devices for which they have been granted access by a customer, through this API.

## Requirements

* Tesla Partner Account - To use this API, you will need to sign up as a Tesla Partner. This requires that you have a name (e.g. sole proprietor or business entity) and website.
* Web Site - You will need to own a domain name (website) and have control of that website for registration and the setup process. 

## Setup

Step 1 - Sign in to Tesla Developer Portal and make an App Access Request: See [Tesla App Access Request](https://developer.tesla.com/request) - During this process, you will need to set up and remember the following account settings:

* CLIENT_ID - This will be provided to you by Tesla when your request is approved.
* CLIENT_SECRET - Same as above.
* DOMAIN - The domain name of a website your own and control.
* REDIRECT_URI - This is the URL that Tesla will direct you to after you authenticate. This landing URL (on your website) will extract the GET variable `code`, which is a one-time use authorization code needed to generate the Bearer auth and Refresh token used to access your Tesla Powerwall energy devices. Place the [index.html](https://github.com/jasonacox/pypowerwall/blob/main/tools/fleetapi/index.html) file in a folder under this domain and use this as the REDIRECT_URI path in the setup below. Alternatively, you can do nothing and just copy the URL from the 404 page during the authorzation process (the code is in the URL). 😁 

Step 2 - Run the [create_pem_key.py](https://github.com/jasonacox/pypowerwall/blob/main/tools/fleetapi/create_pem_key.py) script and place the **public** key on your website at the URL: https://{DOMAIN}/.well-known/appspecific/com.tesla.3p.public-key.pem

Step 3 - Run SETUP using the `fleetapi.py` script. This will ask for all the details above, generate a partner token, register your partner account, and generate a user token needed to access your Powerwall. It will also get the site_id and run a test query to pull live power data for your Powerwall.

```bash
python fleetapi.py setup
```

## Command Line Usage

The `fleetapi.py` script is a command line utility and python class that you can use to monitor and manage your Powerwall. Here are teh command lines:

```
Tesla FleetAPI - Command Line Interface

Usage: fleetapi.py command [arguments] [-h] [--debug] [--config CONFIG] [--site SITE] [--json]

Commands:
    setup               Setup FleetAPI for your site
    sites               List available sites
    status              Report current power status for your site
    info                Display information about your site
    getmode             Get current operational mode setting
    getreserve          Get current battery reserve level setting
    setmode             Set operatinoal mode (self_consumption or autonomous)
    setreserve          Set battery reserve level (prcentage or 'current')
    
options:
  -h, --help            Show this help message and exit
  --debug               Enable debug mode
  --config CONFIG       Specify alternate config file (default: .fleetapi.config)
  --site SITE           Specify site_id
  --json                Output in JSON format
```

Examples

```bash
# Set battery reserve to 80%
python fleetapi.py setreserve 80

# Set battery reserver to be equal to current charge level
python fleetapi.py setreserve current

# Set operatinoal mode to Self Consumption
python fleetapi.py setmode self
```

## FleetAPI Class

You can import and use the FleetAPI class in your own scripts.

```
Class:
    FleetAPI - Tesla FleetAPI Class

 Functions:
    poll(api, action, data) - poll FleetAPI
    get_sites() - get sites
    site_name() - get site name

    get_live_status() - get the current power information for the site
    get_site_info() - get site info
    get_battery_reserve() - get battery reserve level
    get_operating_mode() - get operating mode
    solar_power() - get solar power
    grid_power() - get grid power
    battery_power() - get battery power
    load_power() - get load power
    battery_level() - get battery level
    energy_left() - get energy left
    total_pack_energy() - get total pack energy
    grid_status() - get grid status
    island_status() - get island status
    firmware_version() - get firmware version
    
    set_battery_reserve(reserve) - set battery reserve level (percent)
    set_operating_mode(mode) - set operating mode (self_consumption or autonomous)
     
```

Example Usage

```python
from fleetapi import FleetAPI

fleet = FleetAPI()

# Current Status
print(f"Solar: {fleet.solar_power()}")
print(f"Grid: {fleet.grid_power()}")
print(f"Load: {fleet.load_power()}")
print(f"Battery: {fleet.battery_power()}")
print(f"Battery Level: {fleet.battery_level()}")

# Change Reserve to 30%
fleet.set_battery_reserve(30)

# Change Operating Mode to Autonomous
fleet.set_operating_mode("autonomous")
```

## References

* Developer Documentation about APIs - https://developer.tesla.com/docs/fleet-api#energy-endpoints
* See: https://github.com/jasonacox/pypowerwall/tree/main/tools/fleetapi

**Comment by emptywee** [comment]
@jasonacox so, now that we can plug in different implementations for Tesla API, should we try to add this Fleet API to the mix of the existing ones?

**Reply by jasonacox** [reply]
Hi @emptywee !  Yes, that is on the list!  I'm also considering the [tedapi](https://github.com/jasonacox/Powerwall-Dashboard/discussions/392) one for anyone with a multi-homed setup.

Let me know if you have any thought on how to proceed or would like to help. 100% welcome any help. 😉 🙏

**Reply by emptywee** [reply]
I guess we can start with moving that standalone FleetAPI based tool into its own module just like the cloud and local ones... What do you think?

**Reply by jasonacox** [reply]
I agree. Like cloud, it requires a setup process, but is a bit more elaborate, or at least requires a few more steps/prompt for the user.  But would love to see it part of the library to make it easier.

First blush, I'm thinking...

```bash
# Install pyPowerwall
python3 -m pip install pypowerwall

# (Option 1) By default  will use Local API Option 
python3 -m pypowerwall scan

# (Option 2) Setup to use unofficial Tesla Owners cloud API
python3 -m pypowerwall setup

# (Option 3) Setup to use official Tesla FleetAPI
python3 -m pypowerwall fleetapi
```

```python
# Example Usage

import pypowerwall

# Optional: Turn on Debug Mode
# pypowerwall.set_debug(True)

# (Option 1) Local Mode - Credentials for your Powerwall - Customer Login
password='password'
email='email@example.com'
host = "10.0.1.123"               # Address of your Powerwall Gateway
timezone = "America/Los_Angeles"  # Your local timezone

# (Option 2) Cloud Mode - Requires Setup
password = host = ""
email='email@example.com'
timezone = "America/Los_Angeles"

# (Option 3) FleetAPI Mode - Requires Setup
password = email = host = ""
timezone = "America/Los_Angeles"
client_id = "from_Tesla"               # FleetAPI ID
client_secret = "from_Tesla"        # FleetAPI Secret
domain = "domain"                # Your domain name used for FleetAPI setup

# Connect to Powerwall
pw = pypowerwall.Powerwall(host=host, password=password, email=email, 
             timezone=timezone, client_id=client_id, 
             client_secret=client_secret, domain=domain)
```

**Reply by emptywee** [reply]
I'll begin to work on it in my free time, which I don't have much lately, but we'll see :)

**Reply by jasonacox** [reply]
No worries @emptywee ! Same here. I should have some time in a few weeks to start on it if you don't. Thanks for your help and the kind nudge. :)

**Reply by jasonacox** [reply]
I published a branch: https://github.com/jasonacox/pypowerwall/tree/fleetapi - First hack at it. It's not wired in yet, just roughed in the basics.
